server/sonar-qa-util/src/main/java/org/sonarqube/qa/util/pageobjects/RuleDetails.java

```diff
@@ -19,15 +19,243 @@
+  }
+
+  public RuleDetails shouldHaveType(String type) {
+    $(".coding-rules-detail-property[data-meta=\"type\"]").shouldHave(text(type));
+    return this;
+  }
+
+  public RuleDetails shouldHaveSeverity(String severity) {
+    $(".coding-rules-detail-property[data-meta=\"severity\"]").shouldHave(text(severity));
+    return this;
+  }
+
+  public RuleDetails shouldHaveDescription(String description) {
+    $(".js-rule-description").shouldHave(text(description));
+    return this;
+  }
+
+  public RuleDetails shouldBeActivatedOn(String profileKey) {
+    $("#coding-rules-detail-quality-profiles [data-profile=\"" + profileKey + "\"]").shouldBe(visible);
+    return this;
+  }
+
+  public RuleDetails shouldNotBeActivatedOn(String profileName) {
+    $("#coding-rules-detail-quality-profiles").shouldNotHave(text(profileName));
+    return this;
+  }
+
+  public RuleDetails shouldHaveTotalIssues(int issues) {
+    $(".js-rule-issues h3").shouldHave(text(String.valueOf(issues)));
+    return this;
+  }
+
+  public RuleDetails shouldHaveIssuesOnProject(String projectName, int issues) {
+    $(".coding-rules-most-violated-projects").shouldHave(
+      Condition.and("", text(projectName), text(String.valueOf(issues))));
+    return this;
+  }
+
+  public RuleDetails shouldHaveCustomRule(String ruleKey) {
+    takeCustomRule(ruleKey).shouldBe(visible);
+    return this;
+  }
+
+  public RuleDetails shouldNotHaveCustomRule(String ruleKey) {
+    takeCustomRule(ruleKey).shouldNotBe(visible);
+    return this;
+  }
+
+  public RuleDetails createCustomRule(String ruleName) {
+    $(".js-create-custom-rule").click();
+    modal().shouldBe(visible);
+
+    $("#coding-rules-custom-rule-creation-name").val(ruleName);
+    $("#coding-rules-custom-rule-creation-html-description").val("description");
+    $("#coding-rules-custom-rule-creation-create").click();
+
+    modal().shouldNotBe(visible);
+    return this;
+  }
+
+  public RuleDetails reactivateCustomRule(String ruleName) {
+    $(".js-create-custom-rule").click();
+    modal().shouldBe(visible);
+
+    $("#coding-rules-custom-rule-creation-name").val(ruleName);
+    $("#coding-rules-custom-rule-creation-html-description").val("description");
+    $("#coding-rules-custom-rule-creation-create").click();
+
+    modal().find(".alert-warning").shouldBe(visible);
+    $("#coding-rules-custom-rule-creation-reactivate").click();
+
+    modal().shouldNotBe(visible);
+    return this;
+  }
+
+  public RuleDetails deleteCustomRule(String ruleKey) {
+    takeCustomRule(ruleKey).$(".js-delete-custom-rule").click();
+    modal().shouldBe(visible);
+    modal().find("button").click();
+    modal().shouldNotBe(visible);
+    return this;
+  }
+
+  public RuleActivation activate() {
+    $("#coding-rules-quality-profile-activate").click();
+    modal().shouldBe(visible);
+    return new RuleActivation();
+  }
+
+  private SelenideElement modal() {
+    return $(".modal");
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Define a constant instead of duplicating this literal ".modal" 6 times. ([squid:S1192](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1192))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fstas%2Frules&issues=AWEzbIJCmckxG3vItUxI&open=AWEzbIJCmckxG3vItUxI)

[ISSUE_KEY]: AWEzbIJCmckxG3vItUxI